src/osp/Active/SysRender.cpp

```diff
@@ -187,10 +206,91 @@ void SysRender::display_default_rendertarget(ActiveScene& rScene)
     Renderer::disable(Renderer::Feature::DepthTest);
     Renderer::disable(Renderer::Feature::FaceCulling);
     Renderer::disable(Renderer::Feature::Blending);
+    Renderer::setDepthMask(true);
 
     shader->display_texure(*surface, *target.m_tex);
 }
 
+void SysRender::update_drawfunc_assign(ActiveScene& rScene)
+{
+    ActiveReg_t &rReg = rScene.get_registry();
+    auto &rGroups = rReg.ctx<ACtxRenderGroups>();
+
+    // TODO: check entities that change their material type by checking
+    //       ACompMaterial
+
+    // Loop through each draw group
+    for (auto & [name, rGroup] : rGroups.m_groups)
+    {
+        // Loop through each possibly supported Material ID
+        // from 0 torGroup.m_assigner.size()
+        for (material_id_t i = 0; i < rGroup.m_assigners.size(); i ++)
+        {
+            if (rGroups.m_added.at(i).empty())
+            {
+                continue; // No drawables of this type to assign shaders to
+            }
+
+            RenderGroup::DrawAssigner_t const &assigner
+                    = rGroup.m_assigners.at(i);
+
+            if ( ! bool(assigner) )
+            {
+                continue; // Assigner is null, not set to a function
+            }
+
+            // Pass all new drawables (rGroups.m_added[i]) to the assigner.
+            // This adds accepted entities to rGroup.m_entities
+            assigner(rScene, rGroup.m_entities, rGroups.m_added.at(i));
```

Comment:
looks like m_added and m_assigners are required to be exactly the same size, and that you always iterate them in lock step.

Recommend adding to this rGroup object an iterator type that implements the structured binding protocol. (https://devblogs.microsoft.com/oldnewthing/20201015-00/?p=104369)

Then you can iterate like
```
for(auto & [added, assigner] : rGroup.get_thing_range())
{
...
}
```


which'll make it substantially harder to screw up the lock-step access.

Reply:
I feel that the index has significance in this context, where they are sequential IDs. It's useful to know the index when debugging, and new features may be added that rely on it.